### PR TITLE
[Tools] Add convert tool for Llama models quantized by AutoGPTQ

### PR DESCRIPTION
**Pull request by xiangzez**

### CODE REVIEW

**Comment by Duyi-Wang** [comment]
It's preferable to an additional param for quantification rather than creating a new converter, as this approach allows us to integrate new quantification methods in the future, like AWQ

**Comment by Duyi-Wang** [comment]
And update the description of Converter in README.

**Comment by Duyi-Wang** [comment]
How about `quantization:Optional[str]= "gpqt"`?  `use_gptq` is inconvenient to add new types in future

**Comment by xiangzez** [comment]
For documentation, I think we need a separate tutorial page for quantization. @miaojinc already wrote a [doc](https://github.com/intel/xFasterTransformer/blob/cfabbecc39619464d09e5607787c8a38d7ced15c/tools/quantization/README.md) and we should update this doc and put it in main branch.

**Comment by miaojinc** [comment]
> For documentation, I think we need a separate tutorial page for quantization. @miaojinc already wrote a [doc](https://github.com/intel/xFasterTransformer/blob/cfabbecc39619464d09e5607787c8a38d7ced15c/tools/quantization/README.md) and we should update this doc and put it in main branch.

Sure, I will update the [doc ](https://github.com/intel/xFasterTransformer/blob/feature/awq_qptq/tools/quantization/README.md) after this PR merged.

**Comment by Duyi-Wang** [comment]
> > For documentation, I think we need a separate tutorial page for quantization. @miaojinc already wrote a [doc](https://github.com/intel/xFasterTransformer/blob/cfabbecc39619464d09e5607787c8a38d7ced15c/tools/quantization/README.md) and we should update this doc and put it in main branch.
> 
> Sure, I will update the [doc ](https://github.com/intel/xFasterTransformer/blob/feature/awq_qptq/tools/quantization/README.md) after this PR merged.

@miaojinc Could you add it to our [docs](https://xfastertransformer.readthedocs.io/en/latest/), source code is under [docs branch](https://github.com/intel/xFasterTransformer/tree/docs/docs).

**Comment by miaojinc** [comment]
> > > For documentation, I think we need a separate tutorial page for quantization. @miaojinc already wrote a [doc](https://github.com/intel/xFasterTransformer/blob/cfabbecc39619464d09e5607787c8a38d7ced15c/tools/quantization/README.md) and we should update this doc and put it in main branch.
> > 
> > 
> > Sure, I will update the [doc ](https://github.com/intel/xFasterTransformer/blob/feature/awq_qptq/tools/quantization/README.md) after this PR merged.
> 
> @miaojinc Could you add it to our [docs](https://xfastertransformer.readthedocs.io/en/latest/), source code is under [docs branch](https://github.com/intel/xFasterTransformer/tree/docs/docs).

Yes, sure. I will do that in a new pull request for quantization document.

File: src/xfastertransformer/tools/llama_convert.py [hunk 1]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
+                # qweight is not transposed
+                qweight = param
+                qweight = torch.bitwise_right_shift(torch.unsqueeze(qweight, 1).expand(-1, 32 // self.wbits, -1),
+                        wf.unsqueeze(-1)).to(torch.int16 if self.wbits == 8 else torch.int8)
+                qweight = torch.bitwise_and(qweight, (2 ** self.wbits) - 1)
+                qweight = qweight.reshape(-1, qweight.shape[2])
+                if self.wbits == 8:
+                    qweight = qweight - 128 # uint8 to int8
+                else:
+                    # pack 2 uint4 to 1 int8
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:354]: uint4x2

File: src/xfastertransformer/tools/llama_convert.py [hunk 2]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
+            config["llama"]["layernorm_eps"] = str(hf_config.get("rms_norm_eps", 1e-6))
+            config["llama"]["layernorm_type"] = "pre_layernorm"
+            config["llama"]["activation_type"] = "silu"
+            config["llama"]["has_post_decoder_layernorm"] = "1" if has_post_decoder_layernorm else "0"
+            config["llama"]["vocab_size"] = str(hf_config["vocab_size"])
+            config["llama"]["start_id"] = str(hf_config["bos_token_id"])
+            config["llama"]["end_id"] = str(hf_config["eos_token_id"])
+            config["llama"]["weight_data_type"] = dtype
+
+            config["llama"]["quant_decoder_weights"] = str(True)
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:264]: quant_decoder_weights -> quant_weight_data_type
int8,uint8,int4,uint4,fp8,fp4

File: src/xfastertransformer/tools/convert.py [hunk 3]
```diff
@@ -53,16 +53,22 @@ def get_weight_data_type(self, dtype: str):
         else:
             raise Exception(f"{self.__class__.__name__} don't support convert weight to {dtype}.")
 
-    def convert(self, input_dir, output_dir=None, dtype: str = "fp16", processes=8):
+    def convert(self, input_dir, output_dir=None, dtype: str = "fp16", processes=8, quantization=None):
```
Comment (changqi1) [inline src/xfastertransformer/tools/convert.py:56]: \# from_quantized_model is from HF int8 model to xFT int8 model.
def convert(self, input_dir, output_dir=None, dtype: str = "fp16", processes=8, from_quantized_model=None):

File: src/xfastertransformer/tools/llama_convert.py [hunk 4]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
         pool.close()
         pool.join()
+
+    def split_and_convert_quantized_model(self, input_dir, output_dir, dtype, processes, quantization):
+        """
+        Convert auto-gptq quantized 4 bits or 8 bits Llama model.
+        """
+
+        if quantization != "gptq":
+            print(f"[ERROR] Quantization method {quantization} is not supported.")
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:220]: [ERROR] Please input model must be {quantization} quantized model.

File: src/xfastertransformer/tools/llama_convert.py [hunk 5]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
+            return
+
+        # create directory if not exist
+        if not os.path.exists(output_dir):
+            os.makedirs(output_dir)
+
+        # load the model
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:227]: \# load the quantized model, do not do GPTQ quantization.

File: src/xfastertransformer/tools/llama_convert.py [hunk 6]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
+            print(name)
+
+            if "embed" in name:
+                model_named_parameters[name] = param
+            elif "lm_head" in name:
+                model_named_parameters[name] = param
+            elif "scales" in name:
+                model_named_parameters[name] = param.float()
+            elif "qzeros" in name:
+                qzeros = param
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:333]: zeros is uint8/uint4 from QPTQ quantization, and need to convert to fp32 for xFT.

File: src/xfastertransformer/tools/llama_convert.py [hunk 7]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
+            wf = torch.tensor(list(range(0, 32, self.wbits)), dtype=torch.int32).unsqueeze(0)
+
+            print(name)
+
+            if "embed" in name:
+                model_named_parameters[name] = param
+            elif "lm_head" in name:
+                model_named_parameters[name] = param
+            elif "scales" in name:
+                model_named_parameters[name] = param.float()
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:331]: scales is fp16/fp32/bf16 from QPTQ quantization, and need to convert to fp32 for xFT.

File: src/xfastertransformer/tools/llama_convert.py [hunk 8]
```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
...
+                qzeros = qzeros + 1
+                qzeros = torch.bitwise_and(qzeros, (2 ** self.wbits) - 1)
+                if self.wbits == 8:
+                    qzeros = qzeros - 2 ** (self.wbits - 1) # uint to int
+                qzeros = torch.flatten(qzeros).float()
+                scales = state_dict["model." + name.replace("qzeros", "scales")].float()
+                zeros = - scales * qzeros
+                model_named_parameters[name] = zeros
+            elif "qweight" in name:
+                # qweight is not transposed
```
Comment (changqi1) [inline src/xfastertransformer/tools/llama_convert.py:345]: convert uint8 qweight from QPTQ quantization, and need to convert to int8 for xFT.
convert uint4 qweight from QPTQ quantization, and need to convert to uint4x2 for xFT.

**Review by changqi1 — APPROVED** [review]

**Comment by changqi1** [comment]
@xiangzez CI failed on baichuan model?

**Comment by xiangzez** [comment]
@changqi1 CI issue should be fixed in #287

**Comment by Duyi-Wang** [comment]
> @changqi1 CI issue should be fixed in #287

rebase or merge main branch？

**Comment by changqi1** [comment]
We could rebase to check this PR's status.

**Pull request merged**